integration-tests/src/main/java/org/apache/druid/testing/utils/MsqTestQueryHelper.java

```
@@ -0,0 +1,235 @@
+
+  /**
+   * Submits a {@link SqlQuery} to the MSQ API for execution. This method waits for the task to be accepted by the cluster
+   * and returns the task id associated with the submitted task
+   */
+  public String submitMsqTask(SqlQuery sqlQuery) throws ExecutionException, InterruptedException
+  {
+    String queryUrl = getQueryURL(config.getBrokerUrl());
+    Future<StatusResponseHolder> responseHolderFuture = msqClient.queryAsync(queryUrl, sqlQuery);
+    // It is okay to block here for the result because MSQ tasks return the task Id associated with it, which shouldn't
+    // consume a lot of time
+    StatusResponseHolder statusResponseHolder = responseHolderFuture.get();
+
+    // Check if the task has been accepted successfully
+    HttpResponseStatus httpResponseStatus = statusResponseHolder.getStatus();
+    if (!httpResponseStatus.equals(HttpResponseStatus.ACCEPTED)) {
+      throw new ISE(
+          "Unable to submit the task successfully. Received response status code [%d], and response content:\n[%s]",
+          httpResponseStatus,
+          statusResponseHolder.getContent()
+      );
+    }
+    String content = statusResponseHolder.getContent();
+    SqlTaskStatus sqlTaskStatus;
+    try {
+      sqlTaskStatus = jsonMapper.readValue(content, SqlTaskStatus.class);
+    }
+    catch (JsonProcessingException e) {
+      throw new ISE("Unable to parse the response");
+    }
+    if (sqlTaskStatus.getState().isFailure()) {
+      throw new ISE("Unable to start the task successfully.\nPossible exception: %s", sqlTaskStatus.getError());
+    }
+    return sqlTaskStatus.getTaskId();
+  }
+
+  /**
+   * Polls the overlord API every 1 second and waits for a submitted MSQ task to be completed. Alternatively, one can
+   * specify the maximum time to poll. The method returns the last fetched {@link TaskState} of the task
+   */
+  public TaskState pollTaskIdForCompletion(String taskId, long maxTimeoutSeconds)
```

This piece of code would be already present in some indexing IT no ?

---

If we decide to keep this method, a line mentioning the various timeout cases in the javadocs would be useful.

---

We would never want the IT's to hang so I think we should dis allow "maxTimeouts"
Also what would happen if the taskId is not complete? I feel that the caller of this method would assume the taskID to always complete in the timeout. Should we throw an exception if the task is not completed?

---

Rewritten the code using `RetryUtils` so that it is more error resilient. The only disadvantage that I can think off is the exponential backoff nature of the retries which I was trying to avoid.